tests/test_cases/simulation/test_mpi_ensemble.cu

```diff
@@ -0,0 +1,289 @@
+ *
+ * MPI_Init() and MPI_Finalize() can only be called once each
+ */
+#ifdef FLAMEGPU_ENABLE_MPI
+FLAMEGPU_STEP_FUNCTION(model_step) {
+    int counter = FLAMEGPU->environment.getProperty<int>("counter");
+    counter+=1;
+    FLAMEGPU->environment.setProperty<int>("counter", counter);
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+}
+FLAMEGPU_STEP_FUNCTION(throw_exception) {
+    const int counter = FLAMEGPU->environment.getProperty<int>("counter");
+    const int init_counter = FLAMEGPU->environment.getProperty<int>("init_counter");
+    if (FLAMEGPU->getStepCounter() == 1 && counter == 12) {
+        printf("exception counter: %u, init_counter: %u\n", counter, init_counter);
+        throw std::runtime_error("Exception thrown by host fn throw_exception()");
+    }
+}
+class TestMPIEnsemble : public testing::Test {
+ protected:
+    void SetUp() override {
+        int flag = 0;
+        MPI_Finalized(&flag);
+        if (flag) {
+            // Multi node MPI run
+            GTEST_SKIP() << "Skipping single-node MPI test, MPI_Finalize() has been called";
+            return;
+        }
+        // Init MPI, fetch rank and size
+        int thread_provided = 0;
+        // MPI can only be init once, for certain test cases we do some initial MPI comms for setup
+        MPI_Initialized(&flag);
+        if (!flag) {
+            // MPI single means that only the main thread will perform MPI actions
+            MPI_Init_thread(NULL, NULL, MPI_THREAD_SINGLE, &thread_provided);
+            if (thread_provided != MPI_THREAD_SINGLE) {
+                THROW exception::UnknownInternalError("MPI unable to provide MPI_THREAD_SINGLE support");
+            }
+        }
+        MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+        MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+        // Create a shared memory split
+        MPI_Comm group;
+        MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &group);
+        MPI_Comm_size(group, &group_size);
+        if (world_size < 1) {
+            GTEST_SKIP() << "world_size<1, something went wrong.";
+        }
+        initModel();
+    }
+    void TearDown() override {
+        if (ensemble) delete ensemble;
+        if (exit_log_cfg) delete exit_log_cfg;
+        if (plans) delete plans;
+        if (model) delete model;
+    }
+    void initModel() {
+        model = new flamegpu::ModelDescription("MPITest");
+        model->Environment().newProperty<int>("counter", -1);
+        model->Environment().newProperty<int>("init_counter", -1);
+        model->newAgent("agent");
+        model->newLayer().addHostFunction(model_step);
+    }
+    void initPlans() {
+        // 100 runs, each 10 steps which take >100ms each,
+        // therefore each sim will take >1 second.
+        // Estimate 30 seconds for full test across 4 threads
```

Comment:
Do we need to do this much work in these tests? 5 tests each taking 100s when running with `mpirun -n 1` is a little tedious / is a significant increase in total test suite runtime.

Reply:
Probably not, I suppose I'm just playing safe for poor latency. Probably another argument for splitting MPI tests into distinct suite.

Reply:
This would still be nice to reduce the total test time, 5 tests each taking upto 100s seems excessive. 

Could either reduce the sleep time per test, or reduce the total numeber of tests to be say 10 per MPI rank so it is consistent in time regardless of mpi world size?.

Reply:
I think I prefer the latter idea. Will do